vertical-pod-autoscaler/docs/proposals/crash_loop_handling/README.md

```diff
@@ -0,0 +1,196 @@
+- **exponent**        : A non-zero float exponent value which is determined by the 
+                    targetEstimator, lowedBoundEstimator or upperBoundEstimator. Each of these estimators are set through percentile estimations. targetEstimator is set to percentileEstimator while lowerBoundEstimator and upperBoundEstimator are set as confidence multipliers. For lowerBoundEstimator the exponent is set to -2 and for upperBoundEstimator it is set to +1
+
+## Proposal
+
+To address the concerns mentioned we are proposing this change in the recommendation engine of VPA recommender for autoscaler component.
+
+#### Scale Up:
+Doubling of resources based on current resource usage
+  1. Today we depend on the resource requests history from prometheus client for a history length of 8 days (default)
+  2. Instead, we could obtain current and past 10 minutes usage from prometheus client to get a nearest estimate of how much requests for resources is a pod making
+  3. In the event of a spike in resource, we can recommend the target estimator, requests as double the current resource usage. The lower bound and upper bound become the function of target estimator instead of directly recommending based on resource usage using percentile approach. Lower bound can be set to 0.5 times the target estimator value and upper bound can be set to 2 times the target estimator value.
+  4. This will ensure handling the spike as it is dependent on the current usage than a hysteresis of resource usage pattern and doubling will react exponentially to a potentially exponential spike in resouce usage.
+  5. We consider the following 3 variables (temporary names) threshold_no_of_crashes, last_crash_count_recorded_time, time_length_for_crash_check.
+  6. `threshold_no_of_crashes` talk about value being set during the initiation of recommender to threshold number of restarts due to crash beyond which recommender will scale up both CPU and memory. (Will have a default mostly as 3)
+  7. `last_crash_count_recorded_time` indicates the last crash count time recorded on the node. (Default: 0)
+  8. `time_length_for_crash_check` is `current machine time - last_crash_count_recorded_time`. (Default: 0)
+  9. We can set threshold_no_of_crashes to 3 as the default value. If there is a resource crunch even after bumping one of CPU or memory based on current usage and has resulted in a crash, threshold_no_of_crashes serves a threshold beyond which both resources are to be scaled up irrespective of current usage consumption.
+  10. The variables `threshold_no_of_crashes, last_crash_count_recorded_time, time_length_for_crash_check` are reset to default if the VPA recommender itself restarts. Example: Let's say we recommended the memory doubling but had constant crashes for more than 3 times. Then we get into else condition and double both memory and CPU irrespective of whatever the recommendation for individual resource is.
```

Comment:
I don't fully get the idea behind those variables and how they interact.

The current recommender already watches `OOMKilled` events and add artificially high resource usage samples to its memory histogram. So what I'm asking myself is the following (copied from #47):

> In a discussion [...] we were wondering if it would already help to make [`OOMBumpUpRatio` and `OOMMinBumpUp`](https://github.com/kubernetes/autoscaler/blob/5a7e7cf5cc77c7e7ce5abe3d653a87f295d20248/vertical-pod-autoscaler/pkg/recommender/model/container.go#L29-L32) configurable.
Are you planning on doing so? Or is this not reasonable?


Also we have seen that the recommender sometimes fails to record the aforementioned OOM sample usages, see https://github.com/gardener/gardener/issues/2890#issuecomment-701896155.
Do you have more details under which circumstances this can happen and how this can effect recommendations in case of OOMKills?

Reply:
The above parameters primarily establish a threshold window for scale up and scale down. Other than that, it also ensures we have a crash count to bump both CPU and memory. As said earlier, we have seen cases where we see OOM kills with scale up observed for a particular resource but the other resource is not catching up.

For your second question on observations, I think I will need @amshuman-kr 's help in answering this.